File: djvm/src/main/java/sandbox/java/lang/System.java
package sandbox.java.lang;

@SuppressWarnings({"WeakerAccess", "unused"})
public final class System extends Object {

    private System() {}

    /*
     * This class is duplicated into every sandbox, where everything is single-threaded.
     */
    private static final java.util.Map<java.lang.Integer, java.lang.Integer> objectHashCodes = new java.util.LinkedHashMap<>();
    private static int objectCounter = 0;

    public static int identityHashCode(java.lang.Object obj) {
        int nativeHashCode = java.lang.System.identityHashCode(obj);
        // TODO Instead of using a magic offset below, one could take in a per-context seed
        return objectHashCodes.computeIfAbsent(nativeHashCode, i -> ++objectCounter + 0xfed_c0de);
    }

    public static final String lineSeparator = String.toDJVM("\n");

    public static void arraycopy(java.lang.Object src, int srcPos, java.lang.Object dest, int destPos, int length) {
        java.lang.System.arraycopy(src, srcPos, dest, destPos, length);
    }

    public static void runFinalization() {}
    public static void gc() {}
}

File: djvm/src/main/kotlin/sandbox/java/lang/System.kt
@file:Suppress("UNUSED_PARAMETER", "UNUSED")

package sandbox.java.lang

import java.io.IOException
import java.util.*

object SystemZZZ {

    private var objectCounter = object : ThreadLocal<Int>() {
        override fun initialValue() = 0
    }

    private var objectHashCodes = object : ThreadLocal<MutableMap<Int, Int>>() {
        override fun initialValue() = mutableMapOf<Int, Int>()
    }

    @JvmField
    val `in`: java.io.InputStream? = null

    @JvmField
    val out: java.io.PrintStream? = null

    @JvmField
    val err: java.io.PrintStream? = null

    @JvmStatic
    fun setIn(stream: java.io.InputStream?) {}

    @JvmStatic
    fun setOut(stream: java.io.PrintStream?) {}

    @JvmStatic
    fun setErr(stream: java.io.PrintStream?) {}

    @JvmStatic
    @Throws(IOException::class)
    fun inheritedChannel(): java.nio.channels.Channel? {
        throw UnsupportedOperationException("No inherited channel support")
    }

    @JvmStatic
    fun setSecurityManager(manager: java.lang.SecurityManager?) {}

    @JvmStatic
    fun getSecurityManager(): java.lang.SecurityManager? = null

    @JvmStatic
    fun arraycopy(src: Any?, srcPos: Int, dest: Any?, destPos: Int, length: Int) {
        java.lang.System.arraycopy(src, srcPos, dest, destPos, length)
    }

    @JvmStatic
    fun identityHashCode(obj: Any?): Int {
        val nativeHashCode = java.lang.System.identityHashCode(obj)
        // TODO Instead of using a magic offset below, one could take in a per-context seed
        return objectHashCodes.get().getOrPut(nativeHashCode) {
            val newCounter = objectCounter.get() + 1
            objectCounter.set(newCounter)
            0xfed_c0de + newCounter
        }
    }

    @JvmStatic
    fun getProperties(): java.util.Properties {
        return Properties()
    }

    @JvmStatic
    fun lineSeparator() = "\n"

    @JvmStatic
    fun setProperties(properties: java.util.Properties) {}

    @JvmStatic
    fun getProperty(property: String): String? = null

    @JvmStatic
    fun getProperty(property: String, defaultValue: String?): String? = defaultValue

    @JvmStatic
    fun setProperty(property: String, value: String?): String? = null

    @JvmStatic
    fun clearProperty(property: String): String? = null

    @JvmStatic
    fun getenv(variable: String): String? = null

    @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
    @JvmStatic
    fun getenv(): java.util.Map<String, String>? = null

    @JvmStatic
    fun exit(exitCode: Int) {}

    @JvmStatic
    fun gc() {}

    @JvmStatic
    fun runFinalization() {}

    @JvmStatic
    fun runFinalizersOnExit(flag: Boolean) {}

    @JvmStatic
    fun load(path: String) {}

    @JvmStatic
    fun loadLibrary(path: String) {}

    @JvmStatic
    fun mapLibraryName(path: String): String? = null

}
